File: ClientRuntimes/NodeJS/ms-rest-azure/lib/azureEnvironment.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information. 

'use strict';
var util = require('util');
/**
 * @class
 * Initializes a new instance of the AzureEnvironment class.
 * @constructor
 * @param {string} parameters.name - The Environment name
 * @param {string} parameters.portalUrl - the management portal URL
 * @param {string} parameters.managementEndpointUrl - the management service endpoint
 * @param {string} parameters.resourceManagerEndpointUrl - the resource management endpoint
 * @param {string} parameters.activeDirectoryEndpointUrl - the Active Directory login endpoint
 * @param {string} parameters.activeDirectoryResourceId - The resource ID to obtain AD tokens for
 * @param {string} [parameters.publishingProfileUrl] - the publish settings file URL
 * @param {string} [parameters.sqlManagementEndpointUrl] - the sql server management endpoint for mobile commands
 * @param {string} [parameters.sqlServerHostnameSuffix] - the dns suffix for sql servers
 * @param {string} [parameters.galleryEndpointUrl] - the template gallery endpoint
 * @param {string} [parameters.activeDirectoryGraphResourceId] - the Active Directory resource ID
 * @param {string} [parameters.activeDirectoryGraphApiVersion] - the Active Directory api version
 * @param {string} [parameters.storageEndpointSuffix] - the endpoint suffix for storage accounts
 * @param {string} [parameters.keyVaultDnsSuffix] - the keyvault service dns suffix
 * @param {string} [parameters.azureDataLakeStoreFileSystemEndpointSuffix] - the data lake store filesystem service dns suffix
 * @param {string} [parameters.azureDataLakeAnalyticsCatalogAndJobEndpointSuffix] - the data lake analytics job and catalog service dns suffix
 * @param {bool} [parameters.validateAuthority] - Determines whether the authentication endpoint should 
 * be validated with Azure AD. Default value is true.
 */
function AzureEnvironment(parameters) {
  //Set defaults.
  this.validateAuthority = true;

  if (parameters) {
    //Validate required parameters
    var requiredParams = [ 'name', 'portalUrl', 'managementEndpointUrl', 'resourceManagerEndpointUrl', 
    'activeDirectoryEndpointUrl', 'activeDirectoryResourceId'];
    requiredParams.forEach(function (param) {
      if (!parameters[param] || typeof parameters[param].valueOf() !== 'string') {
        throw new Error(util.format('Please provide "%s" for the environment and it must be of type "string".', param));
      }
    });
    //Assign provided parameters
    for (var prop in parameters) {
      this[prop] = parameters[prop];
    }
  }
}
var supportedEnvironments = {
  Azure: {
    name: 'Azure',
    portalUrl: 'http://go.microsoft.com/fwlink/?LinkId=254433',
    publishingProfileUrl: 'http://go.microsoft.com/fwlink/?LinkId=254432',
    managementEndpointUrl: 'https://management.core.windows.net',
    resourceManagerEndpointUrl: 'https://management.azure.com/',
    sqlManagementEndpointUrl: 'https://management.core.windows.net:8443/',
    sqlServerHostnameSuffix: '.database.windows.net',
    galleryEndpointUrl: 'https://gallery.azure.com/',
    activeDirectoryEndpointUrl: 'https://login.microsoftonline.com/',
    activeDirectoryResourceId: 'https://management.core.windows.net/',
    activeDirectoryGraphResourceId: 'https://graph.windows.net/',
    activeDirectoryGraphApiVersion: '2013-04-05',
    storageEndpointSuffix: '.core.windows.net',
    keyVaultDnsSuffix: '.vault.azure.net',
    azureDataLakeStoreFileSystemEndpointSuffix: 'azuredatalakestore.net',
    azureDataLakeAnalyticsCatalogAndJobEndpointSuffix: 'azuredatalakeanalytics.net'
  },
  AzureChina: {
    name: 'AzureChina',
    portalUrl: 'http://go.microsoft.com/fwlink/?LinkId=301902',
    publishingProfileUrl: 'http://go.microsoft.com/fwlink/?LinkID=301774',
    managementEndpointUrl: 'https://management.core.chinacloudapi.cn',
    resourceManagerEndpointUrl: 'https://management.chinacloudapi.cn',
    sqlManagementEndpointUrl: 'https://management.core.chinacloudapi.cn:8443/',
    sqlServerHostnameSuffix: '.database.chinacloudapi.cn',
    galleryEndpointUrl: 'https://gallery.chinacloudapi.cn/',
    activeDirectoryEndpointUrl: 'https://login.chinacloudapi.cn/',
    activeDirectoryResourceId: 'https://management.core.chinacloudapi.cn/',
    activeDirectoryGraphResourceId: 'https://graph.chinacloudapi.cn/',
    activeDirectoryGraphApiVersion: '2013-04-05',
    storageEndpointSuffix: '.core.chinacloudapi.cn',
    keyVaultDnsSuffix: '.vault.azure.cn',
    // TODO: add dns suffixes for the china cloud for datalake store and datalake analytics once they are defined.
    azureDataLakeStoreFileSystemEndpointSuffix: 'N/A',
    azureDataLakeAnalyticsCatalogAndJobEndpointSuffix: 'N/A'
  },
  AzureUSGovernment: {
    name: 'AzureUSGovernment',
    portalUrl: 'https://manage.windowsazure.us',
    publishingProfileUrl: 'https://manage.windowsazure.us/publishsettings/index',
    managementEndpointUrl: 'https://management.core.usgovcloudapi.net',
    resourceManagerEndpointUrl: 'https://management.usgovcloudapi.net',
    sqlManagementEndpointUrl: 'https://management.core.usgovcloudapi.net:8443/',
    sqlServerHostnameSuffix: '.database.usgovcloudapi.net',
    galleryEndpointUrl: 'https://gallery.usgovcloudapi.net/',
    activeDirectoryEndpointUrl: 'https://login.microsoftonline.com/',
    activeDirectoryResourceId: 'https://management.core.usgovcloudapi.net/',
    activeDirectoryGraphResourceId: 'https://graph.windows.net/',
    activeDirectoryGraphApiVersion: '2013-04-05',
    storageEndpointSuffix: '.core.usgovcloudapi.net',
    keyVaultDnsSuffix: '.vault.usgovcloudapi.net',
    // TODO: add dns suffixes for the US government for datalake store and datalake analytics once they are defined.
    azureDataLakeStoreFileSystemEndpointSuffix: 'N/A',
    azureDataLakeAnalyticsCatalogAndJobEndpointSuffix: 'N/A'
  },
  AzureGermanCloud: {
    name: 'AzureGermanCloud',
    portalUrl: 'http://portal.microsoftazure.de/',
    publishingProfileUrl: 'https://manage.microsoftazure.de/publishsettings/index',
    managementEndpointUrl: 'https://management.core.cloudapi.de',
    resourceManagerEndpointUrl: 'https://management.microsoftazure.de',
    sqlManagementEndpointUrl: 'https://management.core.cloudapi.de:8443/',
    sqlServerHostnameSuffix: '.database.cloudapi.de',
    galleryEndpointUrl: 'https://gallery.cloudapi.de/',
    activeDirectoryEndpointUrl: 'https://login.microsoftonline.de/',
    activeDirectoryResourceId: 'https://management.core.cloudapi.de/',
    activeDirectoryGraphResourceId: 'https://graph.cloudapi.de/',
    activeDirectoryGraphApiVersion: '2013-04-05',
    storageEndpointSuffix: '.core.cloudapi.de',
    keyVaultDnsSuffix: '.vault.microsoftazure.de',
    // TODO: add dns suffixes for the US government for datalake store and datalake analytics once they are defined.
    azureDataLakeStoreFileSystemEndpointSuffix: 'N/A',
    azureDataLakeAnalyticsCatalogAndJobEndpointSuffix: 'N/A'
  }
};

/**
 * Adds a new instance of the AzureEnvironment to the prototype.
 * @param {string} parameters.name - The Environment name
 * @param {string} parameters.portalUrl - the management portal URL
 * @param {string} parameters.managementEndpointUrl - the management service endpoint
 * @param {string} parameters.resourceManagerEndpointUrl - the resource management endpoint
 * @param {string} parameters.activeDirectoryEndpointUrl - the Active Directory login endpoint
 * @param {string} parameters.activeDirectoryResourceId - The resource ID to obtain AD tokens for
 * @param {string} [parameters.publishingProfileUrl] - the publish settings file URL
 * @param {string} [parameters.sqlManagementEndpointUrl] - the sql server management endpoint for mobile commands
 * @param {string} [parameters.sqlServerHostnameSuffix] - the dns suffix for sql servers
 * @param {string} [parameters.galleryEndpointUrl] - the template gallery endpoint
 * @param {string} [parameters.activeDirectoryGraphResourceId] - the Active Directory resource ID
 * @param {string} [parameters.activeDirectoryGraphApiVersion] - the Active Directory api version
 * @param {string} [parameters.storageEndpointSuffix] - the endpoint suffix for storage accounts
 * @param {string} [parameters.keyVaultDnsSuffix] - the keyvault service dns suffix
 * @param {string} [parameters.azureDataLakeStoreFileSystemEndpointSuffix] - the data lake store filesystem service dns suffix
 * @param {string} [parameters.azureDataLakeAnalyticsCatalogAndJobEndpointSuffix] - the data lake analytics job and catalog service dns suffix
 * @param {bool} [parameters.validateAuthority] - Determines whether the authentication endpoint should 
 * be validated with Azure AD. Default value is true.
 * @return {AzureEnvironment} - Reference to the newly added Environment
 */
AzureEnvironment.prototype.add = function(parameters) {
  var _environment = new AzureEnvironment(parameters);
  AzureEnvironment.prototype[_environment.name] = _environment;
  return _environment;
};

//Adding the supported environments
for(var key in supportedEnvironments) {
  AzureEnvironment.prototype.add(supportedEnvironments[key]);
}

module.exports = new AzureEnvironment();

File: src/client/Ruby/ms-rest-azure/lib/ms_rest_azure/azure_environment.rb
# encoding: utf-8
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See License.txt in the project root for license information.

module MsRestAzure
  module AzureEnvironments
    #
    # An instance of this class describes an environment in Azure
    #
    class AzureEnvironment

      # @return [String] the Environment name
      attr_reader :name

      # @return [String] the management portal URL
      attr_reader :portal_url

      # @return [String] the publish settings file URL
      attr_reader :publishing_profile_url

      # @return [String] the management service endpoint
      attr_reader :management_endpoint_url

      # @return [String] the resource management endpoint
      attr_reader :resource_manager_endpoint_url

      # @return [String] the sql server management endpoint for mobile commands
      attr_reader :sql_management_endpoint_url

      # @return [String] the dns suffix for sql servers
      attr_reader :sql_server_hostname_suffix

      # @return [String] the template gallery endpoint
      attr_reader :gallery_endpoint_url

      # @return [String] the Active Directory login endpoint
      attr_reader :active_directory_endpoint_url

      # @return [String] the resource ID to obtain AD tokens for
      attr_reader :active_directory_resource_id

      # @return [String] the Active Directory resource ID
      attr_reader :active_directory_graph_resource_id

      # @return [String] the Active Directory resource ID
      attr_reader :active_directory_graph_api_version

      # @return [String] the endpoint suffix for storage accounts
      attr_reader :storage_endpoint_suffix

      # @return [String] the KeyVault service dns suffix
      attr_reader :key_vault_dns_suffix

      # @return [String] the data lake store filesystem service dns suffix
      attr_reader :datalake_store_filesystem_endpoint_suffix

      # @return [String] the data lake analytics job and catalog service dns suffix
      attr_reader :datalake_analytics_catalog_and_job_endpoint_suffix

      # @return [Boolean] determines whether the authentication endpoint should be validated with Azure AD. Default value is true.
      attr_reader :validate_authority

      def initialize(options)
        required_properties = [:name, :portal_url, :management_endpoint_url, :resource_manager_endpoint_url, :active_directory_endpoint_url, :active_directory_resource_id]

        required_supplied_properties = required_properties & options.keys

        if required_supplied_properties.nil? || required_supplied_properties.empty? || (required_supplied_properties & required_properties) != required_properties
          raise ArgumentError.new("#{required_properties.to_s} are the required properties but provided properties are #{options.to_s}")
        end

        required_supplied_properties.each do |prop|
          if options[prop].nil? || !options[prop].is_a?(String) || options[prop].empty?
            raise ArgumentError.new("Value of the '#{prop}' property must be of type String and non empty.")
          end
        end

        # Setting default to true
        @validate_authority = true

        options.each do |k, v|
          instance_variable_set("@#{k}", v) unless v.nil?
        end
      end
    end

    Azure = AzureEnvironments::AzureEnvironment.new({
                                                        :name => 'Azure',
                                                        :portal_url => 'http://go.microsoft.com/fwlink/?LinkId=254433',
                                                        :publishing_profile_url => 'http://go.microsoft.com/fwlink/?LinkId=254432',
                                                        :management_endpoint_url => 'https://management.core.windows.net',
                                                        :resource_manager_endpoint_url => 'https://management.azure.com/',
                                                        :sql_management_endpoint_url => 'https://management.core.windows.net:8443/',
                                                        :sql_server_hostname_suffix => '.database.windows.net',
                                                        :gallery_endpoint_url => 'https://gallery.azure.com/',
                                                        :active_directory_endpoint_url => 'https://login.microsoftonline.com/',
                                                        :active_directory_resource_id => 'https://management.core.windows.net/',
                                                        :active_directory_graph_resource_id => 'https://graph.windows.net/',
                                                        :active_directory_graph_api_version => '2013-04-05',
                                                        :storage_endpoing_suffix => '.core.windows.net',
                                                        :key_vault_dns_suffix => '.vault.azure.net',
                                                        :datalake_store_filesystem_endpoint_suffix => 'azuredatalakestore.net',
                                                        :datalake_analytics_catalog_and_job_endpoint_suffix => 'azuredatalakeanalytics.net'
                                                    })
    AzureChina = AzureEnvironments::AzureEnvironment.new({
                                                             :name => 'AzureChina',
                                                             :portal_url => 'http://go.microsoft.com/fwlink/?LinkId=301902',
                                                             :publishing_profile_url => 'http://go.microsoft.com/fwlink/?LinkID=301774',
                                                             :management_endpoint_url => 'https://management.core.chinacloudapi.cn',
                                                             :resource_manager_endpoint_url => 'https://management.chinacloudapi.cn',
                                                             :sql_management_endpoint_url => 'https://management.core.chinacloudapi.cn:8443/',
                                                             :sql_server_hostname_suffix => '.database.chinacloudapi.cn',
                                                             :gallery_endpoint_url => 'https://gallery.chinacloudapi.cn/',
                                                             :active_directory_endpoint_url => 'https://login.chinacloudapi.cn/',
                                                             :active_directory_resource_id => 'https://management.core.chinacloudapi.cn/',
                                                             :active_directory_graph_resource_id => 'https://graph.chinacloudapi.cn/',
                                                             :active_directory_graph_api_version => '2013-04-05',
                                                             :storage_endpoing_suffix => '.core.chinacloudapi.cn',
                                                             :key_vault_dns_suffix => '.vault.azure.cn',
                                                             # TODO: add dns suffixes for the china cloud for datalake store and datalake analytics once they are defined.
                                                             :datalake_store_filesystem_endpoint_suffix => 'N/A',
                                                             :datalake_analytics_catalog_and_job_endpoint_suffix => 'N/A'
                                                         })
    AzureUSGovernment = AzureEnvironments::AzureEnvironment.new({
                                                                    :name => 'AzureUSGovernment',
                                                                    :portal_url => 'https://manage.windowsazure.us',
                                                                    :publishing_profile_url => 'https://manage.windowsazure.us/publishsettings/index',
                                                                    :management_endpoint_url => 'https://management.core.usgovcloudapi.net',
                                                                    :resource_manager_endpoint_url => 'https://management.usgovcloudapi.net',
                                                                    :sql_management_endpoint_url => 'https://management.core.usgovcloudapi.net:8443/',
                                                                    :sql_server_hostname_suffix => '.database.usgovcloudapi.net',
                                                                    :gallery_endpoint_url => 'https://gallery.usgovcloudapi.net/',
                                                                    :active_directory_endpoint_url => 'https://login.microsoftonline.com/',
                                                                    :active_directory_resource_id => 'https://management.core.usgovcloudapi.net/',
                                                                    :active_directory_graph_resource_id => 'https://graph.windows.net/',
                                                                    :active_directory_graph_api_version => '2013-04-05',
                                                                    :storage_endpoing_suffix => '.core.usgovcloudapi.net',
                                                                    :key_vault_dns_suffix => '.vault.usgovcloudapi.net',
                                                                    # TODO: add dns suffixes for the US government for datalake store and datalake analytics once they are defined.
                                                                    :datalake_store_filesystem_endpoint_suffix => 'N/A',
                                                                    :datalake_analytics_catalog_and_job_endpoint_suffix => 'N/A'
                                                                })
    AzureGermanCloud = AzureEnvironments::AzureEnvironment.new({
                                                                   :name => 'AzureGermanCloud',
                                                                   :portal_url => 'http://portal.microsoftazure.de/',
                                                                   :publishing_profile_url => 'https://manage.microsoftazure.de/publishsettings/index',
                                                                   :management_endpoint_url => 'https://management.core.cloudapi.de',
                                                                   :resource_manager_endpoint_url => 'https://management.microsoftazure.de',
                                                                   :sql_management_endpoint_url => 'https://management.core.cloudapi.de:8443/',
                                                                   :sql_server_hostname_suffix => '.database.cloudapi.de',
                                                                   :gallery_endpoint_url => 'https://gallery.cloudapi.de/',
                                                                   :active_directory_endpoint_url => 'https://login.microsoftonline.de/',
                                                                   :active_directory_resource_id => 'https://management.core.cloudapi.de/',
                                                                   :active_directory_graph_resource_id => 'https://graph.cloudapi.de/',
                                                                   :active_directory_graph_api_version => '2013-04-05',
                                                                   :storage_endpoing_suffix => '.core.cloudapi.de',
                                                                   :key_vault_dns_suffix => '.vault.microsoftazure.de',
                                                                   # TODO: add dns suffixes for the US government for datalake store and datalake analytics once they are defined.
                                                                   :datalake_store_filesystem_endpoint_suffix => 'N/A',
                                                                   :datalake_analytics_catalog_and_job_endpoint_suffix => 'N/A'
                                                               })
  end
end
